src/devices/Pcx857x/Pcx857x.cs

```diff
@@ -0,0 +1,204 @@
+        public abstract int PinCount { get; }
+
+        public void ClosePin(int pinNumber)
+        {
+            // No-op
+        }
+
+        public void Dispose()
+        {
+            Device.Dispose();
+        }
+
+        public void OpenPin(int pinNumber, PinMode mode) => SetPinMode(pinNumber, mode);
+
+        public PinValue Read(int pinNumber)
+        {
+            Span<PinValuePair> values = stackalloc PinValuePair[]{ new PinValuePair(pinNumber, default) };
+            Read(values);
+            return values[0].PinValue;
+        }
+
+        public void Read(Span<PinValuePair> pinValues)
+        {
+            (uint pins, _) = new PinVector32(pinValues);
+            if (pins >> PinCount > 0)
+                ThrowInvalidPin(nameof(pinValues));
+
+            if ((pins & _pinModes) != 0)
+            {
+                // One of the specified pins was set to output (1)
+                throw new InvalidOperationException("Cannot read from output pins.");
+            }
+
+            ushort data = PinCount == 8 ? ReadByte() : InternalReadUInt16();
+
+            for (int i = 0; i < pinValues.Length; i++)
+            {
+                int pin = pinValues[i].PinNumber;
+                pinValues[i] = new PinValuePair(pin, (data >> pin) & 1);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void ThrowInvalidPin(string argumentName)
+        {
+            // This is a helper to allow the JIT to inline calling methods.
+            // (Methods with throws cannot be inlined.)
+            throw new ArgumentOutOfRangeException(argumentName, $"Pin numbers must be in the range of 0 to {PinCount - 1}.");
+        }
+
+        private void ValidatePinNumber(int pinNumber)
+        {
+            if (pinNumber < 0 || pinNumber >= PinCount)
+                ThrowInvalidPin(nameof(pinNumber));
+        }
+
+        public void SetPinMode(int pinNumber, PinMode mode)
+        {
+            ValidatePinNumber(pinNumber);
+            if (mode == PinMode.Input)
+            {
+                _pinModes = ClearBit(_pinModes, pinNumber);
+            }
+            else if (mode == PinMode.Output)
+            {
+                _pinModes = SetBit(_pinModes, pinNumber);
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException(nameof(mode), "Only Input and Output modes are supported.");
+            }
+
+            WritePins(_pinValues);
+        }
+
+        private void WritePins(ushort value)
+        {
+            // We need to set all input pins to high
+            _pinValues = (ushort)(value | ~_pinModes);
+
+            if (PinCount == 8)
+            {
+                WriteByte((byte)_pinValues);
+            }
+            else
+            {
+                InternalWriteUInt16(_pinValues);
+            }
+        }
+
+        private ushort SetBit(ushort data, int bitNumber) => data |= (ushort)(1 << bitNumber);
```

Comment:
what's the assignment for? Should this be just `data | (ushort)(1 << bitNumber)` or perhaps ref is missing?

Reply:
same comment for ClearBit

Reply:
Nice catch. I've updated this and the Mcp23xxx code where this came from. Also moving the helpers inline as they only have one caller.